qutip/solver/nonmarkov/transfertensor.py

```diff
@@ -0,0 +1,171 @@
+        expectation values or callable or list of callable.
+        Callable signature must be, `f(t: float, state: Qobj)`.
+        See :func:`expect` for more detail of operator expectation.
+
+    options : dictionary
+        Dictionary of options for the solver.
+
+        - store_final_state : bool
+          Whether or not to store the final state of the evolution in the
+          result class.
+        - store_states : bool, None
+          Whether or not to store the state vectors or density matrices.
+          On `None` the states will be saved if no expectation operators are
+          given.
+        - normalize_output : bool
+          Normalize output state to hide ODE numerical errors.
+        - threshold : float
+          Threshold for halting. Halts if  :math:`||T_{n}-T_{n-1}||` is below
+          treshold.
+        - num_learning : int
+          Number of times used to construct the dynmaps operators when
+          ``dynmaps`` is a callable.
+
+    Returns
+    -------
+    output: :class:`qutip.solver.Result`
+        An instance of the class :class:`qutip.solver.Result`.
+
+    .. [1] Javier Cerrillo and Jianshu Cao, Phys. Rev. Lett 112, 110401 (2014)
+    """
+
+    opt = {
+        "store_final_state": False,
+        "store_states": None,
+        "normalize_output": "ket",
+        "threshold": 0.0,
+        "num_learning": 0,
+    }
+    if options:
+        opt.update(options)
+
+    if not np.allclose(np.diff(times), times[1] - times[0]):
+        raise ValueError("The time should be uniformily distributed.")
+
+    if callable(dynmaps):
+        if not options["num_learning"]:
+            raise ValueError(
+                "When dynmaps is a callable, options['num_learning'] must be "
+                "the number of dynamical maps to compute."
+            )
+        dynmaps = [dynmaps(t) for t in times[: opt["num_learning"]]]
+
+    if not dynmaps[0].issuper:
+        raise ValueError("`dynmaps` entries must be super operators.")
+
+    start = time.time()
+    tensors, diff = _generatetensors(dynmaps, opt["threshold"])
+    end = time.time()
+
+    stats = {
+        "preparation time": end - start,
+        "run time": 0.0,
+        "ttmconvergence": diff,
+        "num_tensor": len(tensors),
+    }
+    if state0.isket:
+        state0 = state0.proj()
+
+    if state0.isoper:
+        # vectorize density matrix
+        rho0vec = operator_to_vector(state0)
+        restore = vector_to_operator
+    else:
+        # state0 might be a super in which case we should not vectorize
+        rho0vec = state0
+        restore = lambda state: state
+
+    K = len(tensors)
+    start = time.time()
+    results = Result(e_ops, opt, solver="ttmsolve", stats=stats)
+    states = [rho0vec]
+    results.add(times[0], state0)
+
+    for n in range(1, len(times)):
+        # Set current state
+        state = 0
+        for j in range(1, min(K, n + 1)):
+            tmp = tensors[j] * states[n - j]
```

Comment:
```suggestion
            tmp = tensors[j] @ states[n - j]
```
Should we use `@` here?